sockeye/rerank.py

```diff
@@ -41,36 +41,69 @@ class Reranker:
     :param return_score: If True, also return the sentence-level score.
     """
 
-    def __init__(self, metric: str,
-                 return_score: bool = False) -> None:
-        if metric == C.RERANK_BLEU:
+    def __init__(self, metric: str, return_score: bool = False) -> None:
+        self.metric = metric
+        self.return_score = return_score
+
+        if self.metric == C.RERANK_BLEU:
             # "add-k" smoothing is the best-performing method implemented in
             # sacrebleu.  See "Method 2" results from Chen and Cherry
             # (http://aclweb.org/anthology/W14-3346)
             self.scoring_function = partial(sacrebleu.sentence_bleu, smooth_method='add-k')
-        elif metric == C.RERANK_CHRF:
+        elif self.metric == C.RERANK_CHRF:
             self.scoring_function = sacrebleu.sentence_chrf  # type: ignore
+        elif self.metric.startswith(C.RERANK_ISOMETRIC):
+            self.scoring_function = utils.compute_isometric_score  # type: ignore
         else:
             raise utils.SockeyeError("Scoring metric '%s' unknown. Choices are: %s" % (metric, C.RERANK_METRICS))
 
-        self.return_score = return_score
 
-    def rerank(self, hypotheses: Dict[str, Any], reference: str) -> Dict[str, Any]:
+    def rerank(self, hypotheses: Dict[str, Any], reference: str, source: str,
+               isometric_alpha: float = 0.5) -> Dict[str, Any]:
         """
         Reranks a set of hypotheses that belong to one single reference
         translation. Uses stable sorting.
 
         :param hypotheses: Nbest translations.
         :param reference: A single string with the actual reference translation.
+        :param source: A single string with the source sentence.
+        :param isometric_alpha: For re-ranking using isometric translation criterion.
         :return: Nbest translations sorted by reranking scores.
         """
-        scores = [self.scoring_function(hypothesis, [reference]).score for hypothesis in hypotheses['translations']]
-        ranking = list(np.argsort(scores, kind='mergesort')[::-1])  # descending
+
+        if self.metric == C.RERANK_BLEU or self.metric == C.RERANK_CHRF:
+            scores = [self.scoring_function(hypothesis, [reference]).score for
+                      hypothesis in hypotheses['translations']]
+            # BLEU,CHRF - the higher, the better
+            ranking = self._get_ranking_indices(scores, kind='mergesort', descending=True)
+
+        if self.metric.startswith(C.RERANK_ISOMETRIC):
+            scores = [self.scoring_function(hypothesis, hypothesis_score[0], source,
+                                            self.metric, isometric_alpha)
```

Comment:
if you add `isometric_alpha` to the constructor, you could create a partial `scoring_function` with constant arguments, alpha & metric, in the constructor, and only pass hypothesis, score, source here.